### PR TITLE
SpaceAPI Client as a real Module

### PR DESCRIPTION
**Pull request by dns2utf8**
This contains the WIP for a SpaceAPI Client as a Module so it can be used in `main()`

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -51,14 +52,40 @@ fn main() {
...
+                                Some(nth) => { try!(send(&api, m, format!("invalid OptionalInteger: {}", nth))); return Ok(ListeningAction::Continue); },*/
+                                Some (nth) => match w.get(nth as usize) {
+                                    Some(e) => {},
+                                    None => {},
+                                },
+                                None => {},
+                            };
+                            
+                            for pic_path in w {
+                                let caption = sac.basename(&pic_path);
```
Comment (dbrgn) [inline src/main.rs:90]: Are you sure we need to add a caption? :)

File: src/spaceapi_client.rs [hunk 2]
```diff
@@ -1,84 +1,203 @@
...
+}
+
+impl SpaceApiClient {
+  pub fn new() -> SpaceApiClient {
+    let mut s = SpaceApiClient{
+      last_fetch: 0,
+      webcams: Vec::new(),
+    };
+    
+    s.fetch_webcams();
```
Comment (dbrgn) [inline src/spaceapi_client.rs:31]: I'm not sure if the constructor should have I/O side effects. Maybe move this to an `init()` method?

File: src/spaceapi_client.rs [hunk 3]
```diff
@@ -1,84 +1,203 @@
...
+  
+  pub fn fetch_people_now_present(&self) -> ::std::result::Result<u64, String> {
+    let status = try!(fetch_status());
+    
+    extract_people_now_present(status)
+  }
+
+  pub fn get_webcams(&self) -> Vec<String> {
+    self.webcams.clone()
+  }
```
Comment (dbrgn) [inline src/spaceapi_client.rs:44]: Why not make the `webcams` attribute public? Getters sound a bit Java-ish. It can also be confused with `fetch_webcams`.

File: src/spaceapi_client.rs [hunk 4]
```diff
@@ -1,84 +1,203 @@
...
+    
+    for c in path.chars() {
+      match c {
+        '/' => r.clear(),
+        _ => r.push( c ),
+      }
+    }
+    
+    r
+  }
```
Comment (dbrgn) [inline src/spaceapi_client.rs:80]: This is quite an inefficient approach to calculating the basename, you have a lot of unnecessary read/write operations.

This would be much nicer using a lazy reverse iterator.

But in the end, it's not needed if you use [`Path.file_name()`](https://doc.rust-lang.org/std/path/struct.Path.html#method.file_name) :)

File: src/spaceapi_client.rs [hunk 5]
```diff
@@ -1,84 +1,203 @@
...
 }
 
+/// Fetch a Binary from url and save it to a temporary Location.
+/// returns the temp Path
+fn fetch_binary(url :&String) -> Result<Vec<u8>,io::Error> {
+  let client = Client::new();
+  
+  let mut res = match client.get(url).send() {
+    Ok(v) => v,
+    Err(e) => return Err(io::Error::new(io::ErrorKind::BrokenPipe, format!("{:?}", e))),
```
Comment (dbrgn) [inline src/spaceapi_client.rs:133]: Is this always a broken pipe error?

File: src/spaceapi_client.rs [hunk 6]
```diff
@@ -1,84 +1,203 @@
...
+  };
+  
+  let mut v = vec![];
+  let size = try!(res.read_to_end(&mut v));
```
Comment (dbrgn) [inline src/spaceapi_client.rs:137]: Shouldn't binary data be stored in a byte array?

File: src/spaceapi_client.rs (hunk 4 above)
Comment (rnestler) [inline src/spaceapi_client.rs:80]: Yes don't reinvent the wheel, use the std lib.

File: src/spaceapi_client.rs (hunk 3 above)
Comment (dns2utf8) [inline src/spaceapi_client.rs:44]: because I will modify it late and I prefer to hand over a Clone

File: src/spaceapi_client.rs (hunk 4 above)
Comment (dns2utf8) [inline src/spaceapi_client.rs:80]: Does it work with `http://...`?

File: src/spaceapi_client.rs (hunk 5 above)
Comment (dns2utf8) [inline src/spaceapi_client.rs:133]: No but I must change the Error-Type and this is the closest one

File: src/spaceapi_client.rs (hunk 6 above)
Comment (dns2utf8) [inline src/spaceapi_client.rs:137]: It is :wink: `Vec<u8>`

File: src/spaceapi_client.rs (hunk 3 above)
Comment (dbrgn) [inline src/spaceapi_client.rs:44]: But the user can still do `sac.webcams.clone()`.

If you want to keep it this way, at least I'd call the function `clone_webcams`. Right now it's not really obvious what it does.

File: src/spaceapi_client.rs (hunk 4 above)
Comment (dbrgn) [inline src/spaceapi_client.rs:80]: Ah, you want to use it on URLs. Neither the function name nor the parameter name (path) nor the documentation explains that :)

In that case, what about this implementation?

``` rust
/// Return the last part of a path or URL.                                         
fn basename<'a>(path: &'a str) -> &'a str {
    match path.rfind('/') {
        Some(p) => &path[p+1..],
        None => &path,
    }
}

#[cfg(test)]
mod tests {
    use super::basename;

    #[test]
    fn path() {
        assert_eq!(basename("/usr/bin/bash"), "bash");
    }

    #[test]
    fn url() {
        assert_eq!(basename("http://coredump.ch/f/index.html"), "index.html");
    }

    #[test]
    fn plain() {
        assert_eq!(basename("foobar"), "foobar");
    }

    #[test]
    fn none() {
        assert_eq!(basename("/bin/"), "");
    }
}
```

That searches for the slash from the right side. Also, it works with slices, so you simply return a new slice into the original string, without allocating a new one.

Comment (dbrgn) [inline src/spaceapi_client.rs:80]: Hm, looks like you can even leave away the lifetime annotations:

``` rust
/// Return the last part of a path or URL.                                         
fn basename(path: &str) -> &str {
    match path.rfind('/') {
        Some(p) => &path[p+1..],
        None => &path,
    }
}
```

File: src/spaceapi_client.rs (hunk 5 above)
Comment (dbrgn) [inline src/spaceapi_client.rs:133]: The ErrorKind is optional, as far as I know :) Or at least you can set it to something like `Other`.

File: src/spaceapi_client.rs (hunk 6 above)
Comment (dbrgn) [inline src/spaceapi_client.rs:137]: Yeah, I should probably have read the docs :)

File: src/spaceapi_client.rs (hunk 4 above)
Comment (rnestler) [inline src/spaceapi_client.rs:80]: Yes, if there is only one parameter the lifetime gets inferred, see https://doc.rust-lang.org/book/lifetimes.html#lifetime-elision

> If there is exactly one input lifetime, elided or not, that lifetime is assigned to all elided lifetimes in the return values of that function.

Comment (rnestler) [inline src/spaceapi_client.rs:80]: Did you take a look at the [url crate](https://crates.io/crates/url)?

Comment (dbrgn) [inline src/spaceapi_client.rs:80]: That doesn't seem to offer a method that does this. OTOH you could convert an `Url` to a `Path` and then use the `file_name` method.

File: src/spaceapi_client.rs [hunk 7]
```diff
@@ -1,84 +1,231 @@
...
+    
+    f.write_all(&bin);
+    f.sync_all();
+    
+    Ok::<String,io::Error>(format!("{}", path.to_str().unwrap()))
+  }
+  
+  pub fn basename(&self, path :&String) -> String {
+    match path.rfind('/') {
+      Some(p) => String::from( &path[p+1..] ),
```
Comment (dbrgn) [inline src/spaceapi_client.rs]: Hm, why did you change back to `&String` and `String` again? :( That's an unneeded heap allocation, a slice is much more flexible.

Comment (rnestler) [inline src/spaceapi_client.rs]: Yeah the solution from @dbrgn looked better to me:

``` Rust
fn basename(path: & str) -> & str {
    match path.rfind('/') {
        Some(p) => &path[p+1..],
        None => &path,
    }
}
```

Comment (dns2utf8) [inline src/spaceapi_client.rs]: I take BabySteps, plus I need it as a String anyways.

**Comment by dns2utf8** [comment]
I had to update all crates, waiting for upstream to merge https://github.com/LukasKalbertodt/telegram-bot/pull/34

**Comment by dbrgn** [comment]
The commit history is now quite messy, I lost the overview. Just merge, and in the future, try to avoid merge commits in pull requests :)

File: src/main.rs [hunk 8]
```diff
@@ -63,72 +62,88 @@ fn main() {
...
+                                        return Ok(ListeningAction::Continue);
+                                    } else {
+                                        let mut n = 0;
+                                        cams.filter(|_| {
+                                            n == nth
+                                        })
+                                    }
+                                },
+                                None => {
+                                    cams.filter(|_| { true })
```
Comment (dns2utf8) [inline src/main.rs]: I am having trouble figuring out how to compile this. :stuck_out_tongue_closed_eyes: 
I would like to filter one element sometimes, but somtimes not.
